src/validation/recursive_validator.rs

```diff
@@ -0,0 +1,123 @@
+use std::cell::Cell;
+
+use indexmap::{IndexMap, IndexSet};
+
+use crate::{diagnostics::Diagnostic, index::Index, typesystem::DataTypeInformationProvider};
+
+/// Status of whether a node has been visited or not.
+#[derive(Clone, Copy, Hash, PartialEq, Eq)]
+enum Status {
+    Visited,
+    Unvisited,
+}
+
+/// Validator to find and report all recursive data structures using Depth-first search (DFS)[1].
+/// Such data structures consists of structs and function-blocks, for example the following code would be
+/// flagged as a recursive data structure:
+/// ```st
+/// TYPE A : STRUCT
+///     b : B;
+/// END_STRUCT END_TYPE
+///
+/// TYPE B : STRUCT
+///     a : A;
+/// END_STRUCT END_TYPE
+/// ```
+/// Without a validator the compiler would panic because generating `A -> B -> A -> ...` is impossible as it
+/// overflows the stack.
+///
+/// [1] https://en.wikipedia.org/wiki/Depth-first_search
+pub struct RecursiveValidator {
+    pub diagnostics: Vec<Diagnostic>,
+}
+
+impl RecursiveValidator {
+    pub fn new() -> RecursiveValidator {
+        RecursiveValidator { diagnostics: Vec::new() }
+    }
+
+    /// Entry point of finding and reporting all recursive data structures.
+    pub fn validate_recursion(&mut self, index: &Index) {
+        let structs = index
+            .get_types()
+            .values()
+            .filter(|x| x.get_type_information().is_struct())
+            .map(|x| (x.get_name(), Cell::new(Status::Unvisited)))
+            .collect();
+
+        let fbs = index
+            .get_pous()
+            .values()
+            .filter(|x| x.is_function_block())
+            .map(|x| (x.get_name(), Cell::new(Status::Unvisited)))
+            .collect();
+
+        for nodes in vec![structs, fbs] {
```

Comment:
I believe this introduces a problem because the nodes in find_cycles either contain all structs or all FBs, so the check in line 88 may not find the status if for example a FB has a field of type Struct and vice versa. 

So a cycle between structs and FBs will not be detected. 
```
TYPE A : STRUCT
    b : B;
END_STRUCT END_TYPE

FUNCTION_BLOCK B 
    VAR_INPUT
        a : A;
    END_VAR
END_FUNCTION_BLOCK
```